[WIP, RFC FS-1072] task and state machine support

dsyme opened the pull request:
[updated]

This inserts a heavily modified (but semantically almost completely compatible) version of [TaskBuilder.fs](https://github.com/rspeele/TaskBuilder.fs) into FSharp.Core and adds a general state machine compilation mechanism for F# computation expressions.

### Overview

The primary intention is to add quality support for `task { ... }` to F#.  This means 

* [x] library support for Task `task { ... }`
* [x] state machine compilation for computation expressions
* [x] support for using task-like tasks (value tasks etc.)
* [x] support for configurable tasks (half done)
* [ ] support for defining value tasks 
* [ ] tests for all of these (half done)
* [ ] proper debug support (stack, breakpoints, stepping) equivalent to C# (ok but needs testing)

We won't add the support until all of the above are done.  We are starting with TaskBuilder.fs as a reference library implementation to help define semantics.

The mechanism we use to do this is to first add general support for "generated state machines" in F#. 

### Technical Note: Specifying generated state machines 

From a high level, state machines are lovely to implement inefficiently in a functional language e.g. using continuations or F# `async { ... }` or TaskBuilder.fs `task { ... }`. These tend to allocate continuations and other values like crazy.  Thus their **efficient** implementation (with low or zero allocation rates) means generating C-like constructs - label, goto, jumptables and and other such things.  

It is hard to recover a low-allocation implementation from a functional/continuation/... implementation directly - but we can if we add enough compiler support for generating exactly the right code.  Generation of efficient state machines from needs compiler support. For example both C# and F# support state machine compilation of C# iterator methods and F# `seq { .... }` expressions.  

The magic heart of a typical generated state machine is a `MoveNext` or `Step` function that takes an integer `program counter` (`pc`) and jumps to a target:

```fsharp
      member __.MoveNext() = 
            match (pc) with
               | 1 -> goto L1 
               | 2 -> goto L2 
               | _ -> goto L0

            L0: ...
                ... this code can return, first setting "pc <- L1"...

            L1: ...
                ... this code can return, e.g. first setting "pc <- L2"...

            L2: ...

```
This is roughly what compiled `seq { ... }` code looks like in F# today and what compiled async/await code looks like in C#, at a very high level.

Note you can't write this kind of code directly in F# - there is no `goto` and the `goto` often jump directly into code resuming from the last step of the state machine.

In this mechanism we allow the specification of new varieties of generated state machines in **library** code, normally as part of the implementation of a computation-expression builder.  (Note this is an extremely subtle mechanism and its validity is not yet checked by the compiler - "caveat emptor, here be dragons").

To help define generated state machines we use some primitives, currently
```fsharp
module Microsoft.FSharp.Core.CompilerServices.CodeGenHelpers
        val __jumptable : int -> (unit -> 'T) -> 'T
        val __stateMachine : 'T -> 'T
        val __newEntryPoint: unit -> int
        val __machine<'T> : 'T
        val __entryPoint: int -> unit
        val __return : 'T -> 'T
```
plus some special magic value names such `__expand_XYZ` and the very obscure `__machine_step$cont`.

A generated state machine expression is any expression of the form 
```fsharp
    let inline makeStateMachine __expand_code = 
        (__stateMachine
            { new SomeBaseType() with 
                member __.SomeOverride(pc) =
                     __jumptable pc __expand_code }).SomeMethod()
```
Here `SomeOverride` will be compiled as a multi-entry method where the entry point is determined by `pc` using a jumptable at the start of the method.  

**Important** The content of a generated  state machine is specified by `__expand_code` which **must** be fully inlined code - that is, fully inlined to reveal the full implementation of the state machine. You can think of everything beginning with `__expand_ABC` as a macro where macro expansion is implemented by F# inlining.

For example, `__expand_code` could be a call to :
```
    makeStateMachine (bindTask task (fun res -> returnTask (res + 1))
```
where `bindTask` is inlined:
```fsharp
    let inline bindTask (task : Task<'TResult1>) (__expand_continuation : 'TResult1 -> TaskStep<'TResult2>) =
        let CONT = __newEntryPoint()
        let awaiter = task.GetAwaiter()
        if awaiter.IsCompleted then 
            __entryPoint CONT
            __expand_continuation (awaiter.GetResult())
        else
            __machine<TaskStateMachine>.Await (awaiter, CONT)
            __return false
```
and `returnTask` is inlined:
```fsharp
    let inline returnTask (x: 'T)  =
        __machine<TaskStateMachine>.Current <- x
        true
```

This shows the use of some of the other constructs in state machine specification - you can see 
1. the compile-time creation of an entry point using `__newEntryPoint()`
2. the compile-time specification of the target of that entry point using `__entryPoint`
3. the compulsory macro expansion of the continuation using `__expand_continuation`
4. a self-reference to the state machine being run (using the `this` pointer) via `__machine`
5. an early exit from the state machine stepping function using `__return`

The constructs that can be used in the (inlined, expanded) state machine code is limited and in some cases (e.g. try/with blocks and `while` loops) extremely subtle.  It is easy to create incorrect and invalid code using this mechanism.  

For full details of current status see the implementation in the PR, details may have changed from the above.

When specifying state machines, it is common to return a typed "dummy" struct such as `TaskStep<'T>` from each call to the `Step` function where `T` represents the result of the task.  This is sort of a phantom type.


### Example: `sync { ... }`

As a micro "no-op" example of defining a builder which gets compiled using state machines, we can define `sync { ... }` which is for entirely synchronous computation with no special semantics.

Implementation: https://github.com/dsyme/visualfsharp/blob/tasks/tests/fsharp/core/state-machines/sync.fs

Examples of use:
```fsharp
let t1 y = 
    sync {
       printfn "in t1"
       let x = 4 + 5 + y
       return x
    }

let t2 y = 
    sync {
       printfn "in t2"
       let! x = t1 y
       return x + y
    }

printfn "t2 6 = %d" (t2 6)
```
Code performance will be approximately the same as normal F# code except for one allocation for each execution of each `sync { .. }` as we allocate the "SyncMachine".  In later work we may be able to remove this.

### Example: `task { ... }`

See the implementation in [tasks.fs](https://github.com/dsyme/visualfsharp/blob/tasks/tests/fsharp/core/state-machines/list.fshttps://github.com/microsoft/visualfsharp/pull/6634/files#diff-71685d8abf5330863b6d92402f9132fbR211).  There is complication due to the need to bind to task-pattern tasks and asyncs.

### Example: `taskSeq { ... }`

This is for state machine compilation of computation expressions that generate `IAsyncEnumerable<'T>` values. This is a headline C# 8.0 feature and a very large feature for C#.  It appears to mostly drop out as library code once general-purpose state machine support is available.

See the example in [taskSeq.fs](https://github.com/dsyme/visualfsharp/blob/tasks/tests/fsharp/core/state-machines/taskSeq.fs).  Not everything is implemented yet but the basics work.

### Example `seq2 { ... }`

See https://github.com/dsyme/visualfsharp/blob/tasks/tests/fsharp/core/state-machines/seq2.fs

This is an example showing how to do state machine compilation for `seq2 { ... }` expressions, akin to `seq { ... }` expressions, for which we bake-in state machine compilation into the F# compiler today. Caveats:

* It doesn't cover tailcalls for recursive sequence generation
* I haven't checked the disposal logic, though I think it's broadly correct
* No perf testing done (it may be better, may be worse)
https://github.com/microsoft/visualfsharp/pull/6634/files#diff-4837d60671e85e130108370a6a8c0597R169

I think it's possible this version actually gives better stack traces than the current sequence expression support in the F# compiler.

This is essentially trimming the task support out of `taskSeq { ... }` .

### Examples `list { ... }`, `array { ... }`

See https://github.com/dsyme/visualfsharp/blob/tasks/tests/fsharp/core/state-machines/list.fs

This example defines  `list { .. }`, `array { .. }` and `rsarray { .. }` for collections, where the computations generate directly into a `ResizeArray` (`System.Collections.Generic.List<'T>`).

F#'s existing `[ .. ]` and `[| ... |]` and `seq { .. } |> Seq.toResizeArray` all use an intermediate `IEnumerable` which is then iterated to populate a `ResizeArray` and then converted to the final immutable collection. In contrast, generating directly into a `ResizeArray` is potentially more efficient (and for `list { ... }` further perf improvements are possible if we put this in `FSharp.Core` and use the mutate-tail-cons-cell trick to generate the list directly). This technique has been known for a while and can give faster collection generation but it has not been possible to get good code generation for the expressions in many cases. Note these aren't really "state machines" because there are no resumption points - there is just an implicit collection we are yielding into in otherwise synchronous code.

Using a directly-generating `list { ... }` seems to give a significant speedup over `[ ... ]` in the example I just tried, included in the code.

```
PERF: list { ... } : 497
PERF: [ ... ] : 748
```

### Technical Note: Expected allocation profile for `task { ... }`

The allocation performance of the current approach should be:

* [x] one allocation of TaskStateMachine per `task { ... }`
* [x] one allocation of Task per `task { ... }`
* [x] one or two allocations on each `let!` or `do!` bind in a `task { .. }` - I'm not quite sure how many (we may be able to remove these when binding to another task produced by `task { ... }`, though I'm not sure).
* [x] there is an additional boxing allocation when a bind returns a value type - we can remove this later
one allocation on each let mutable used inside the task - these currently get turned into ref cells through * [x] the autobox transformation when `let mutable` is used in a task

More improvements are needed - see discussion below. We should compare with TaskBuilder.fs, Ply and C#.

### Performance Status

Systematic perf testing of `task { ... }` is required.

Some benchmarks are at `tests\fsharp\perf\tasks` in the PR. Please help improve this.  

Currently compile and run with:

```
msbuild tests\fsharp\perf\tasks\FS\TaskPerf.fsproj /p:Configuration=Release
dotnet artifacts\bin\TaskPerf\Release\netcoreapp2.1\TaskPerf.dll
```

The build/run  cycle is a bit irritating as BenchmarkDotNet seems to run the "FSharpAsync" slow benchmarks around 100 times. Please help fix that.

Here are results at last run:
```
BenchmarkDotNet=v0.11.5, OS=Windows 10.0.18362
Intel Core i7-8750H CPU 2.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=2.2.203
  [Host]     : .NET Core 2.1.9 (CoreCLR 4.6.27414.06, CoreFX 4.6.27415.01), 64bit RyuJIT DEBUG
  DefaultJob : .NET Core 2.1.9 (CoreCLR 4.6.27414.06, CoreFX 4.6.27415.01), 64bit RyuJIT
```

|                    Method |     Mean |     Error |    StdDev |   Median | Ratio | RatioSD |     Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------- |---------:|----------:|----------:|---------:|------:|--------:|----------:|------:|------:|----------:|
| ManyWriteFile_CSharpAsync | 18.19 ms | 0.6308 ms | 1.8499 ms | 17.49 ms |  1.00 |    0.00 |  406.2500 |     - |     - |   1.09 KB |
|        ManyWriteFile_Task | 16.66 ms | 0.3535 ms | 0.9062 ms | 16.46 ms |  0.92 |    0.09 |  406.2500 |     - |     - |   1.02 KB |
| ManyWriteFile_TaskBuilder | 20.79 ms | 0.4120 ms | 1.0411 ms | 20.83 ms |  1.16 |    0.11 | 1125.0000 |     - |     - |   1.75 KB |
| ManyWriteFile_FSharpAsync | 32.59 ms | 0.6645 ms | 1.9489 ms | 32.80 ms |  1.81 |    0.18 | 1750.0000 |     - |     - |   3.46 KB |

|                Method |      Mean |    Error |   StdDev |    Median | Ratio | RatioSD |       Gen 0 | Gen 1 | Gen 2 |  Allocated |
|---------------------- |----------:|---------:|---------:|----------:|------:|--------:|------------:|------:|------:|-----------:|
| SyncBinds_CSharpAsync |  94.21 ms | 1.942 ms | 4.872 ms |  92.28 ms |  1.00 |    0.00 | 167833.3333 |     - |     - |  755.31 MB |
|        SyncBinds_Task | 120.77 ms | 1.322 ms | 1.104 ms | 120.72 ms |  1.26 |    0.09 | 167800.0000 |     - |     - |  755.31 MB |
| SyncBinds_TaskBuilder | 160.14 ms | 2.100 ms | 1.964 ms | 160.30 ms |  1.66 |    0.11 | 256000.0000 |     - |     - | 1152.04 MB |
| SyncBinds_FSharpAsync | 864.69 ms | 3.520 ms | 2.748 ms | 865.17 ms |  9.07 |    0.55 | 596000.0000 |     - |     - | 2685.55 MB |

|                 Method |      Mean |    Error |   StdDev | Ratio | RatioSD |     Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|---------:|---------:|------:|--------:|----------:|------:|------:|----------:|
| AsyncBinds_CSharpAsync | 121.53 ms | 2.416 ms | 3.901 ms |  1.00 |    0.00 | 1000.0000 |     - |     - |   1.87 MB |
|        AsyncBinds_Task |  94.16 ms | 1.820 ms | 2.235 ms |  0.77 |    0.03 | 1200.0000 |     - |     - |   2.33 MB |
| AsyncBinds_TaskBuilder | 123.48 ms | 2.234 ms | 2.090 ms |  1.01 |    0.04 | 3600.0000 |     - |     - |   4.96 MB |

|                     Method |        Mean |      Error |     StdDev | Ratio | RatioSD |       Gen 0 | Gen 1 | Gen 2 |    Allocated |
|--------------------------- |------------:|-----------:|-----------:|------:|--------:|------------:|------:|------:|-------------:|
| SingleSyncTask_CSharpAsync |    59.46 ms |  0.3397 ms |  0.3011 ms |  1.00 |    0.00 |           - |     - |     - |            - |
|        SingleSyncTask_Task |    63.98 ms |  0.5184 ms |  0.4849 ms |  1.08 |    0.01 |           - |     - |     - |            - |
| SingleSyncTask_TaskBuilder |    94.40 ms |  1.2734 ms |  1.1911 ms |  1.59 |    0.02 | 127000.0000 |     - |     - |  600000000 B |
| SingleSyncTask_FSharpAsync | 1,160.43 ms | 15.3916 ms | 14.3973 ms | 19.54 |    0.24 | 813000.0000 |     - |     - | 3840000000 B |

dsyme commented:
Note the library implementation is not big - just 300 lines.  And it's pretty transparent what is happening apart from the advanced use of SRTP resolution

dsyme commented:
The main problem here is adding the signature file, so large chunks are commented out.

cartermp commented:
Link suggestion since the RFC isn't out yet: https://github.com/fsharp/fslang-suggestions/issues/581

dsyme commented:
I have updated this 

* The implementation and signature files now match and are complete (though untested)

* I have hidden the representation of `TaskStep` and made it a struct-around-reference-type wrapper, both to allow future improvements in representation and future reduction of allocations

* various renamings

* changed the use of `obj`/`IComparable`/... pattern for giving overload resolution priority order to a new `IPriority1`, `IPriority2`, `IPriority3` pattern with those types in `FSharp.Core.CompilerServices`

dsyme commented:
cc @gusty , @rspeele

eiriktsarpalis commented:
Looks interesting, but I'm wondering if support for cancellation could be somehow incorporated in this. I know it doesn't compose the same way as async does, but I mostly use task builders to interface with F# agnostic libraries: `task { }` for the outer expressions and `async { }` for anything else. In such scenaria cancellation in task builders does work rather well (c.f. http://fssnip.net/7TE/title/Easy-task-creation line 48 for an example).

ashtonkj commented:
@dsyme - Will this be compatible with Orleans (https://github.com/dotnet/orleans). I know Taskbuilder.fs is, but Ply isn't (as the continuations aren't guaranteed to run on the same thread breaking the Orleans threading model). I'm hoping if this is using the same base as Taskbuilder.fs it will be.

dsyme commented:
> @dsyme - Will this be compatible with Orleans (https://github.com/dotnet/orleans). I know Taskbuilder.fs is, but Ply isn't (as the continuations aren't guaranteed to run on the same thread breaking the Orleans threading model). I'm hoping if this is using the same base as Taskbuilder.fs it will be.

Yes, as things stand it is semantically identical to TaskBuilder.fs

gusty commented on src/fsharp/FSharp.Core/tasks.fs:
```diff
@@ -0,0 +1,391 @@
+
+namespace Microsoft.FSharp.Core.CompilerServices
+    /// A marker interface to give priority to different available overloads
+    type IPriority3 = interface end
```
Just curious, why an interface in lieu of a class?
In F#+ I use `... class end`.
Then in Witnesses I would just inherit from Priority3.

dsyme replied:
No great preference either way

dsyme commented:
I'm exploring state machine generation.  The rough approach is 
1. [Hack in a pseudo-code implementation of state machines into the existing PR](https://github.com/Microsoft/visualfsharp/compare/master...dsyme:tasks2?expand=1).  Primitives are `__jsr`, `__genlabel` and `__gencode`
2. Consider an approach where we inline all code and have the F# compiler recognise these primitives as a way of generating state machines

OnurGumus commented:
Also consider supporting async main methods for being on par with C#

cartermp commented:
@onurgumus that is out of scope for this PR/feature. Can you open a language suggestion if one doesn’t exist? Thanks!

OnurGumus commented:
@cartermp I disagree with it being out of scope. To me it seems to objective of this feature is to achieve C# compatibility with respect to Tasks. If that is the case,  main functions that can return a Task should be a part of it.  Of course from implementation point of view it maybe difficult. But that's another story.

cartermp commented:
@OnurGumus Can you open a language suggestion if one doesn't exist?

dsyme commented:
I've been trying to make progress on this.  I have a new branch based on a careful transformation of TaskBuilder.fs   https://github.com/Microsoft/visualfsharp/compare/master...dsyme:tasks3?expand=1  where 

1. Each builder combinator has body that is essentially the state machine codegen we must perform, expressed in terms of a set of codegen primitives `__newLabel`, `__label`, `__goto` and so on 
2. The codegen primitives match the primitives that already exist in the F# TAST and used for sequence state machine compilation 
3. If the codegen were actually implemented, then the semantics would, i believe, be correct (it is a careful transformation of the original TaskBuilder.fs).
4. If the codegen were actually implemented on the fully inlined code, then the performance would, I believe, be 
   - one allocation per `task { ... }`
   - one or two allocations on each `let!` bind on a task - I'm not quite sure how many - we may be able to remove these when binding to another task produced by `task { ... }`, though I'm not sure.
   - there is an additional boxing allocation when a bind returns a value type - we can remove this later
   - one allocation on each `let mutable` used inside the task - these currently get turned into ref cells through the autobox transformation (IIRC this is also true in state machine compilation)

   I'm pretty sure this would be better than TaskBuilder today, I'm not sure how it would compare to `ply` though.  Also it doesn't deal with making `ValueTask`

Note the branch above deliberately doesn't execute correctly since the codegen primitives are mocked for now.

This is all very tricky/detailed work... ugh....

OnurGumus commented:
@dsyme item 6 would be proper debugging and stack traces perhaps.

dsyme commented:
> @dsyme item 6 would be proper debugging and stack traces perhaps.


Thanks, added that to the list

NatElkins commented:
Is the intention to have the same behavior as in C#?  So that if you do something like this:

```fsharp
let tsk = task { return 1 }
```

the task is "hot" and already running?

dsyme commented:
> Is the intention to have the same behavior as in C#? So that if you do something like this:

Yes.  As I understand things this is the same as TaskBuilder.fs.

NinoFloris commented:
@ashtonkj Ply not compatible with Orleans? 
It should be as it's context sensitive so it should resume back via any synch context that was set during yield. Could you [open an issue](https://github.com/crowded/ply) if you have more details?

@dsyme out of scope right now but some great enhancements over C# async await would be:
- Shipping a task CE that thunks its body and returns a cold/unstarted `Task<TResult>(Func<Task<TResult>>` (re-usability is still subject to Task only completing once) `lazyTask {}` ?
- Can CEs accept args? If so have `task {}` accept a CancellationToken that is checked after every resumption `token.ThrowIfCancellationRequested()` by the state machine. It won't flow etcetera but it's better than manual checking.
- Easily detach from captured context like `task.ConfigureAwait(continueOnCapturedContext = false)` but for the full CE scope instead. TaskBuilder has two versions of `task {}` ContextSensitive `ConfigureAwait(true)` and ContextInsensitive `ConfigureAwait(false)`. Something like that is super useful to have as a builtin.

Some general information about the allocation profile a C# Task/ValueTask method would have. 
This is on .NET Core 2.2+

There are 0 allocations when completing synchronously for `ValueTask` if our return type is `Task`, we have to allocate that one obviously.
When going async however we get:
- A box of the state machine — this happens inside [AsyncMethodBuilder](https://github.com/dotnet/coreclr/blob/027a9105ae2783a7fffd6038bf07a5cf67ba287b/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs) — that box is also a derived `Task`/`Task<TResult>` instance
- A `MoveNext()` delegate for that box to pass to all `awaiter.OnCompleted(action: Action)` calls which proxies a call to the original state machine's `MoveNext()` method.

These are all framework allocations, effectively the generated state machine does not allocate anything.

dsyme commented:
For visibility I'm going to start pushing the state machine generation version of this, though it's very far from working

For reference, and for use as a comparison semantics, the last functioning commit that was a port of TaskBuilder.fs is here: https://github.com/Microsoft/visualfsharp/pull/6634/commits/4185ce6d529db6fadef2c3802f5895e2b26560df

dsyme commented:
Good news everyone, state machine compilation of tasks now works, in the sense that all TaskBuilder.fs tests now pass.  There are a few restrictions like no `let rec` within task bodies.

There is a lot of work left to do but I think this is getting close where we could start to try some performance tests.

I've updated the description at the top of this PR to reflect latest status and also included a description of the state machine mechanism this introduces.  If you're a deep-hacker you might like to try to use the mechanism before we restrict its use to FSharp.Core (assuming we do that!) :)

dsyme commented:
For the deep F# computation expression hackers out there, I've added an example of a `sync { ... }` computation expression that compiles using the state machine mechanism above

Neftedollar commented:
I've found that I can't use protected methods inside classes with taskBuilder in the computation expressions. As I understand it's cause computation expressions are sugar for lambdas and I can't use protected methods from lambdas. This implementation still with CE and with this implementation I still can't use protected methods inside of task CE. Am I right?

benaadams commented:
> A `MoveNext()` delegate for that box to pass to all awaiter.OnCompleted(action: Action) calls which proxies a call to the original state machine's MoveNext() method.

The C# version uses [private interfaces `ITaskAwaiter` and `IConfiguredTaskAwaiter`](https://github.com/dotnet/coreclr/blob/48431cc037776ca359de36bf71bda8c154cc2aa9/src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs#L399-L438) to call `TaskAwaiter.UnsafeOnCompletedInternal` which additionally passes the statemachine in so MoveNext() can be called without the delegate needing to be allocated.

NinoFloris commented:
Right that type check and consequent internal magic is done for AwaitUnsafeCompleted. AwaitCompleted (not sure how frequently that's still used) directly grabs the action.

Doesn't matter either way though as this implementation makes use of the same AsyncMethodbuilder(s) as C# does, so all optimizations are included!

forki commented:
@dsyme at least at some point in the past (not sure if now) the TaskBuilder.fs had this automatic conversion from async to task in `let!` - will this version have it?

(please say "no")

Tarmil commented:
> (please say "no")

I assume you're worried that people would be unknowingly hit by the overhead of the conversion? I guess it would be easy to mistakenly call `let! x = AsyncFoo()` instead of `let! x = FooAsync()` and it would work, just slower than it should.

forki commented:
No not really worried about overhead. More about change in semantics that should be made explicit.

NinoFloris commented:
With trait extensions we could possibly leave this off. Anybody that wants it could easily add an extension member for it. If marked inline it should be able to participate in the rewriting. (but that hasn't been tested yet)

dsyme commented:
> @dsyme at least at some point in the past (not sure if now) the TaskBuilder.fs had this automatic conversion from async to task in let! - will this version have it?

It is present in the prototype since it is present in TaskBuilder.fs.   The main concern relates to the cancellation token of course.

dsyme commented:
To show the power of the general-purpose state machine generation, I've added an example implementing state machine conversion for IAsyncEnumerable specified by `taskSeq { ... }`.

You can find the builder and the examples of its usage here: https://github.com/microsoft/visualfsharp/pull/6634/files#diff-c484d5dd4d3de5b84fa6695e3aeb868dR225

Note this is a "massive" feature in C# 8.0/9.0.   With general-purpose support for state machines a preliminary version of this took a few hours.  (nb. there are many more details to be sorted out of course - and we have not yet checked/refined debugging and perf ).

dsyme commented:
And I've added another example showing how to do state machine compilation for sequences, something we bake into the F# compiler today.  Caveats: 
* It doesn't cover tailcalls for recursive sequence generation
* I haven't checked the disposal logic, though I think it's broadly correct
* No perf testing done (it may be better, may be worse)

https://github.com/microsoft/visualfsharp/pull/6634/files#diff-4837d60671e85e130108370a6a8c0597R169

I think it's possible this version actually gives better stack traces than the current sequence expression support in the F# compiler.  

That took 25 minutes to implement, simply trimming the task support out of the `taskSeq { ... }` example....

baronfel commented:
@dsyme what's the development story like for these? is it possible still to use the `Quote` member to validate your CE call-chains?

dsyme commented:
> @dsyme what's the development story like for these? is it possible still to use the Quote member to validate your CE call-chains?

Well, quotations will return the call chains.  However it won't reveal the state machines

dsyme commented:
And finally I added an example `list { .. }`, `array { .. }` and `rsarray { .. }` for state machine compilation of sequence expressions that generate directly into a ResizeArray (System.Collections.Generic.List<'T>).

The existing `[ .. ]` and `[| ... |]` and `seq { .. } |> Seq.toResizeArray` all use an intermediate `IEnumerable` which is then iterated.  In contrast, generating directly into a ResizeArray is potentially more efficient. This technique has been known for a while and can give faster collection generation but it has not been possible to get state machine compilation for the expressions.

Using state-machine compiled `list { ... }` instead of `[ ... ]` gives a significant speedup in the example I just tried:
```
PERF: list { ... } : 497
PERF: [ ... ] : 748
```

For `list { ... }` further perf improvements are possible if we put this in FSharp.Core and use the mutate-tail-cons-cell trick to generate the list

OnurGumus commented:
Perhaps at this point we should also check with Alfonso regarding Fable compatibility.

dsyme commented:
@OnurGumus good point. Fable will see the inlining (since it runs the Optimizer.fs phase) - but not the final state machine conversion. So some kind of support in Fable compiler would be needed or else these CEs will simply fail.

forki commented:
AFAIK fable is not running the optimizer yet. We tried to push in that direction but we did not succeed yet. @alfonsogcnunez @ncave that's still correct right?

ncave commented:
> Fable is not running the optimizer yet. We tried to push in that direction but we did not succeed yet.

@forki No, that's not entirely correct anymore, since we closed most optimizer issues. The optimizer is fully functional now, but running it is optional (behind a compiler option).

forki commented:
So it's working? Cool. But is it enabled by default?

ncave <notifications@github.com> schrieb am So., 12. Mai 2019, 14:35:

> Fable is not running the optimizer yet. We tried to push in that direction
> but we did not succeed yet.
>
> @forki <https://github.com/forki> No, that's not entirely correct
> anymore, since we closed most optimizer issues. The optimizer is fully
> functional now, but running it is optional (behind a compiler option).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/visualfsharp/pull/6634#issuecomment-491591853>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAOAND4QLOPPEEXD57I66DPVAFJ3ANCNFSM4HIG4VKQ>
> .
>

ncave commented:
@forki The optimizer is fully functional but not enabled by default (can be enabled with a compiler option).

dsyme commented:
I did a quick performance test using the read/write IO benchmark in TaskBuilder.fs repository. This is promising, it shows performance about equal to C# for this particular benchmark.
```
C# methods completed in 2232 ms
task { .. } completed in 2201 ms
TaskBuilder task { .. } completed in 2413 ms
F# async completed in 5644 ms
F# async (AwaitTask) completed in 4462 ms
```
then 
```
C# methods completed in 2185 ms
task { .. } completed in 2220 ms
TaskBuilder task { .. } completed in 2396 ms
F# async completed in 5781 ms
F# async (AwaitTask) completed in 4432 ms
```
The benchmark is dominated by repeated calls to `WriteAsync`: 
```
            for i = 1 to writeIterations do
                do! file.WriteAsync(junk, 0, junk.Length)
```

cartermp commented:
@dsyme Would it be possible to run those as BenchmarkDotNet benchmarks? Single runs like this could very well look like they're fine but have a large error range. Additionally, using the `MemoryDiagnoser` attribute will give a sense for what allocations look like.

dsyme commented:
> @dsyme Would it be possible to run those as BenchmarkDotNet benchmarks? Single runs like this could very well look like they're fine but have a large error range. Additionally, using the MemoryDiagnoser attribute will give a sense for what allocations look like.

Yeah eventually. But it's accurate enough for now TBH - the results tend to be steady within +/- 5%.  Someone really needs to do a very detailed perf analysis (and it's generally better that it's not the person doing the core feature otherwise things slip through...)

dsyme commented:
I added a different variation of the perf test where `WriteAsync` is replaced by `Task.FromResult`, i.e. testing the degenerate case where there's not really anything async happening, but the compiler can't really tell this without special knowledge of `Task.FromResult`.

In this case, somewhat as expected, the current `task { .. }` implementation improves compare to TaskBuilder.fs but is slower than C# (which has additional support to reduce allocations on entirely synchronous paths).


```
-------- operation = WRITE_ASYNC ------
C# methods completed in 1144 ms
task { .. } completed in 1198 ms
TaskBuilder task { .. } completed in 1399 ms
F# async completed in 3060 ms
F# async (AwaitTask) completed in 2349 ms
-------- operation = FROM_RESULT ------
C# methods completed in 173 ms
task { .. } completed in 675 ms
TaskBuilder task { .. } completed in 803 ms
F# async completed in 5573 ms
F# async (AwaitTask) completed in 5635 ms
```

I've updated the description at the top with a section on performance.  It would be great if someone could take `tests\fsharp\perf\tasks` and turn it into a full performance test suite using BenchmarkDotnet, and including Ply etc.

cartermp commented:
@dsyme See here: https://github.com/microsoft/visualfsharp/issues/6718

dsyme commented:
I improved the perf benchmark suite and added performance results in the top PR description.

In `ManyAsyncWrites` we're doing fine:

|                    Method |        Mean |      Error |      StdDev |      Median |
|-------------------------- |------------:|-----------:|------------:|------------:|
| ManyWriteFile_CSharpAsync |    19.05 ms |  0.3721 ms |   0.4967 ms |    19.08 ms |
|        ManyWriteFile_Task |    18.81 ms |  0.3739 ms |   0.7552 ms |    18.82 ms |
| ManyWriteFile_TaskBuilder |    20.39 ms |  0.3947 ms |   0.4848 ms |    20.55 ms |

The biggest of all outlier is `SingleSyncTask` Here the `task { ... }` support is slower than TaskBuilder.fs.  This is kind of expected at the moment as we allocate a state machine and method builder even when the task completes synchronously, will work on that.

|                    Method |        Mean |      Error |      StdDev |      Median |
|-------------------------- |------------:|-----------:|------------:|------------:|
|    SingleSyncTask_CSharpAsync |    58.27 ms |  0.3168 ms |   0.2808 ms |    58.25 ms |
|           SingleSyncTask_Task |   189.00 ms |  3.5567 ms |   2.9700 ms |   189.30 ms |
|    SingleSyncTask_TaskBuilder |    98.36 ms |  2.3028 ms |   3.2281 ms |    97.08 ms |

cartermp commented:
@dsyme FYI I sent you a PR that does benchmarks a bit more cleanly and also adds memory analysis.

dsyme commented:
Note to run the perf tests you need to use:
```
msbuild tests\fsharp\perf\tasks\FS\TaskPerf.fsproj /p:Configuration=Release
dotnet artifacts\bin\TaskPerf\Release\netcoreapp2.1\TaskPerf.dll

```

dsyme commented:
I've implemented the (very difficult) job of using value type representations for task state machines.

This is AFAIK the last "hard" bit of the performance story w.r.t. allocation (except allowing the definition of ValueTask which is not yet in scope).  

The perf results are now below.  There's still some extra overhead in the F# `SyncBinds` and `SingleSyncTask` that I'm not quite sure about (though it's less than before). The figures are now very close to C# for the other two benchmarks, within noise really.


|                    Method |     Mean |     Error |    StdDev | Ratio | RatioSD |
|-------------------------- |---------:|----------:|----------:|------:|--------:|
| ManyWriteFile_CSharpAsync | 17.31 ms | 0.2648 ms | 0.2477 ms |  1.00 |    0.00 |
|        ManyWriteFile_Task | 17.24 ms | 0.3395 ms | 0.7379 ms |  1.02 |    0.04 |
| ManyWriteFile_TaskBuilder | 19.37 ms | 0.3850 ms | 0.5994 ms |  1.13 |    0.04 |
| ManyWriteFile_FSharpAsync | 28.57 ms | 0.5659 ms | 1.4808 ms |  1.71 |    0.10 |

|                Method |      Mean |     Error |    StdDev | Ratio | RatioSD |
|---------------------- |----------:|----------:|----------:|------:|--------:|
| SyncBinds_CSharpAsync |  90.87 ms |  1.701 ms |  1.508 ms |  1.00 |    0.00 |
|        SyncBinds_Task | 126.56 ms |  1.926 ms |  1.708 ms |  1.39 |    0.03 |
| SyncBinds_TaskBuilder | 158.97 ms |  3.012 ms |  2.818 ms |  1.75 |    0.03 |
| SyncBinds_FSharpAsync | 903.88 ms | 22.757 ms | 30.380 ms |  9.83 |    0.18 |

|                 Method |     Mean |    Error |   StdDev | Ratio | RatioSD |
|----------------------- |---------:|---------:|---------:|------:|--------:|
| AsyncBinds_CSharpAsync | 115.7 ms | 1.535 ms | 1.436 ms |  1.00 |    0.00 |
|        AsyncBinds_Task | 122.7 ms | 2.393 ms | 3.726 ms |  1.07 |    0.04 |
| AsyncBinds_TaskBuilder | 162.4 ms | 3.292 ms | 9.232 ms |  1.36 |    0.04 |

|                     Method |        Mean |     Error |    StdDev | Ratio | RatioSD |
|--------------------------- |------------:|----------:|----------:|------:|--------:|
| SingleSyncTask_CSharpAsync |    63.67 ms |  1.235 ms |  1.423 ms |  1.00 |    0.00 |
|        SingleSyncTask_Task |   120.55 ms |  2.291 ms |  3.634 ms |  1.89 |    0.07 |
| SingleSyncTask_TaskBuilder |    99.16 ms |  1.998 ms |  3.991 ms |  1.58 |    0.08 |
| SingleSyncTask_FSharpAsync | 1,148.98 ms | 21.611 ms | 20.215 ms | 18.13 |    0.57 |

cartermp commented:
@dsyme Looks like you missed the `[<MemoryDiagnoser>]` attribute for each benchmark. That will give us a look into allocations per op and GC stats in a perf run.

Arnavion commented on src/fsharp/FSharp.Core/tasks.fs:
```diff
@@ -0,0 +1,432 @@
+        let __newEntryPoint() : int = failwith "__newEntryPoint should always be removed from compiled code"
+
+        [<MethodImpl(MethodImplOptions.NoInlining)>]
+        let __machine<'T> : 'T = failwith "__newEntryPoint should always be removed from compiled code"
```
Should the error message mention `__machine` instead of `__newEntryPoint` ?

dsyme commented:
I have updated this PR removing one source of allocation, and updated the performance results

The allocation profile of the F# code is now identical to that of the C# baselines, except for `AsyncBinds` where there are some extra allocations (though fewer than TaskBuilder).  However this doesn't hugely concern me as truly async binding almost always incurs other signficant allocations or delays. THe important thing is that there are now no extra allocations for synchronous binding.

The main thing now is to consider how to make the state machine feature sufficiently complete to allow its incorporation.  From the examples I've worked through I'm convinced of the general utility of the mechanism, however there are some kinds of code that can't currently be compiled to state machines.

NinoFloris commented on src/fsharp/FSharp.Core/tasks.fs:
```diff
@@ -0,0 +1,424 @@
+                  = bindTask (Async.StartAsTask computation) __expand_continuation     
+
+        // Give the type arguments explicitly to make it match the signature precisely
+        static member inline CanReturnFrom< ^TaskLike, ^Awaiter, ^T
```
It seems indeed ReturnFrom (the let bound function) isn't used for CanReturnFrom

NinoFloris commented on src/fsharp/FSharp.Core/tasks.fs:
```diff
@@ -0,0 +1,424 @@
+            TaskStep<'TResult2, 'TOverall>(false)
+
+// New style task builder.
+type TaskBuilder() =
```
We could we make this a value type right? Otherwise sync completion is still an alloc for the TaskBuilder

NinoFloris commented on src/fsharp/FSharp.Core/tasks.fs:
```diff
@@ -0,0 +1,424 @@
+    /// trying to decide between satisfying the constraints with `Task` or `Task<'TResult1>`.
+    let inline bindTask (task : Task<'TResult1>) (__expand_continuation : 'TResult1 -> TaskStep<'TResult2, 'TOverall>) =
+        let CONT = __newEntryPoint()
+        let awaiter = task.GetAwaiter()
```
Same, awaiter could be a mutable struct.

NinoFloris commented on src/fsharp/FSharp.Core/tasks.fs:
```diff
@@ -0,0 +1,424 @@
+                                            and ^Awaiter : (member get_IsCompleted : unit -> bool)
+                                            and ^Awaiter : (member GetResult : unit -> ^TResult1) >
+            (awaitable : ^Awaitable, __expand_continuation : ^TResult1 -> TaskStep<'TResult2, 'TOverall>) : TaskStep<'TResult2, 'TOverall> =
+                let awaiter = (^Awaitable : (member GetAwaiter : unit -> ^Awaiter)(awaitable)) // get an awaiter from the awaitable
```
I'm not sure if all awaiter implementations are immutable structs. Might want to mark it mutable.

C# stores local awaiter on statemachine after calling IsCompleted, allowing mutations during that call.
Note that I haven't actually seen a mutable awaiter ever, at least not in the BCL.

In any case it doesn't hurt to remove the copy for calling its methods.

dsyme replied:
This awaiter goes on the state machine because it is used after `CONT`.  I've made it mutable, thanks!

NinoFloris commented on src/fsharp/FSharp.Core/tasks.fs:
```diff
@@ -0,0 +1,424 @@
+
+    let inline Await (sm: byref<TaskStateMachineTemplate<'T>>) (awaiter, pc) =
+        sm.ResumptionPoint <- pc
+        let mutable awaiter = awaiter
```
Should be able to remove this if awaiter is passed byref, see later comments

dsyme replied:
I think it's not possible - the method expects a `byref` not `inref`

NinoFloris replied:
A byref argument isn't an inref though?

dsyme replied:
Done now, thanks!

NinoFloris commented on src/fsharp/FSharp.Core/tasks.fs:
```diff
@@ -0,0 +1,424 @@
+                                                        and ^Awaiter : (member get_IsCompleted : unit -> bool)
+                                                        and ^Awaiter : (member GetResult : unit -> ^TResult1) >
+            (task : ^TaskLike, __expand_continuation : ^TResult1 -> TaskStep<'TResult2, 'TOverall>) : TaskStep<'TResult2, 'TOverall> =
+                let awaitable = (^TaskLike : (member ConfigureAwait : bool -> ^Awaitable)(task, false))
```
Same, awaiter could be a mutable struct.

NinoFloris commented on src/fsharp/FSharp.Core/tasks.fs:
```diff
@@ -0,0 +1,424 @@
+    /// defined on the builder has fancy generic constraints on inp and T parameters.
+    let inline bindTaskConfigureFalse (task : Task<'TResult1>) (__expand_continuation : 'TResult1 -> TaskStep<'TResult2, 'TOverall>) =
+        let CONT = __newEntryPoint ()
+        let awaiter = task.ConfigureAwait(false).GetAwaiter()
```
Same, awaiter could be a mutable struct.

dsyme commented on src/fsharp/FSharp.Core/tasks.fs:
```diff
@@ -0,0 +1,424 @@
+        // A using statement is just a try/finally with the finally block disposing if non-null.
+        builder.TryFinally(
+            (fun () -> __expand_body disp),
+            (fun () -> if not (isNull (box disp)) then disp.Dispose()))
```
How is this done? My understanding is `Object.ReferenceEquals(disp, null)` will also box, unless the JIT elmiinates?

NinoFloris replied:
Jit eliminates due to generic specialization via the flexibile type sig.

dsyme replied:
I think we should double check either way

NinoFloris replied:
Definitely. Though it's quite a simple path, boxes are semantics changing so cannot be eliminated easily while ReferenceEquals gets aggressively inlined to what amounts to `cmp disp, null; jne` which for 'T = struct always is `false = true`. Branch elimination does the rest.

dsyme commented:
@NinoFloris Thanks for the review.  I've applied the changes and fixed a race condition.

Latest perf results are shown below  F# is 26% slower for SyncBinds, and 23% faster than C# for AsyncBinds, at least on this test run. Allocations are zero for repeated SyncBindSingleTask, which is as hoped.


|                    Method |     Mean |     Error |    StdDev |   Median | Ratio | RatioSD |     Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------- |---------:|----------:|----------:|---------:|------:|--------:|----------:|------:|------:|----------:|
| ManyWriteFile_CSharpAsync | 18.19 ms | 0.6308 ms | 1.8499 ms | 17.49 ms |  1.00 |    0.00 |  406.2500 |     - |     - |   1.09 KB |
|        ManyWriteFile_Task | 16.66 ms | 0.3535 ms | 0.9062 ms | 16.46 ms |  0.92 |    0.09 |  406.2500 |     - |     - |   1.02 KB |
| ManyWriteFile_TaskBuilder | 20.79 ms | 0.4120 ms | 1.0411 ms | 20.83 ms |  1.16 |    0.11 | 1125.0000 |     - |     - |   1.75 KB |
| ManyWriteFile_FSharpAsync | 32.59 ms | 0.6645 ms | 1.9489 ms | 32.80 ms |  1.81 |    0.18 | 1750.0000 |     - |     - |   3.46 KB |

|                Method |      Mean |    Error |   StdDev |    Median | Ratio | RatioSD |       Gen 0 | Gen 1 | Gen 2 |  Allocated |
|---------------------- |----------:|---------:|---------:|----------:|------:|--------:|------------:|------:|------:|-----------:|
| SyncBinds_CSharpAsync |  94.21 ms | 1.942 ms | 4.872 ms |  92.28 ms |  1.00 |    0.00 | 167833.3333 |     - |     - |  755.31 MB |
|        SyncBinds_Task | 120.77 ms | 1.322 ms | 1.104 ms | 120.72 ms |  1.26 |    0.09 | 167800.0000 |     - |     - |  755.31 MB |
| SyncBinds_TaskBuilder | 160.14 ms | 2.100 ms | 1.964 ms | 160.30 ms |  1.66 |    0.11 | 256000.0000 |     - |     - | 1152.04 MB |
| SyncBinds_FSharpAsync | 864.69 ms | 3.520 ms | 2.748 ms | 865.17 ms |  9.07 |    0.55 | 596000.0000 |     - |     - | 2685.55 MB |

|                 Method |      Mean |    Error |   StdDev | Ratio | RatioSD |     Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|---------:|---------:|------:|--------:|----------:|------:|------:|----------:|
| AsyncBinds_CSharpAsync | 121.53 ms | 2.416 ms | 3.901 ms |  1.00 |    0.00 | 1000.0000 |     - |     - |   1.87 MB |
|        AsyncBinds_Task |  94.16 ms | 1.820 ms | 2.235 ms |  0.77 |    0.03 | 1200.0000 |     - |     - |   2.33 MB |
| AsyncBinds_TaskBuilder | 123.48 ms | 2.234 ms | 2.090 ms |  1.01 |    0.04 | 3600.0000 |     - |     - |   4.96 MB |

|                     Method |        Mean |      Error |     StdDev | Ratio | RatioSD |       Gen 0 | Gen 1 | Gen 2 |    Allocated |
|--------------------------- |------------:|-----------:|-----------:|------:|--------:|------------:|------:|------:|-------------:|
| SingleSyncTask_CSharpAsync |    59.46 ms |  0.3397 ms |  0.3011 ms |  1.00 |    0.00 |           - |     - |     - |            - |
|        SingleSyncTask_Task |    63.98 ms |  0.5184 ms |  0.4849 ms |  1.08 |    0.01 |           - |     - |     - |            - |
| SingleSyncTask_TaskBuilder |    94.40 ms |  1.2734 ms |  1.1911 ms |  1.59 |    0.02 | 127000.0000 |     - |     - |  600000000 B |
| SingleSyncTask_FSharpAsync | 1,160.43 ms | 15.3916 ms | 14.3973 ms | 19.54 |    0.24 | 813000.0000 |     - |     - | 3840000000 B |

NinoFloris commented:
The asyncbinds result makes me suspicious though. Why is it deviating so much from both C# and Taskbuilder? (which are quite close to each other). Could it point to a behavioral difference?

Good to see a full MB of allocations, about 30%, shaved off by those changes, leaving just a tiny difference. Really have to see the IL for that last bit though, could be an FSharpRef or some other hiding in there.

dsyme commented:
> The asyncbinds result makes me suspicious though. Why is it deviating so much from both C# and Taskbuilder? (which are quite close to each other). Could it point to a behavioral difference?

I'll do a few more runs tomorrow, the results are a bit variable. I suspect the C# one just came in slow on that particular test run 

> Good to see a full MB of allocations, about 30%, shaved off by those changes, leaving just a tiny difference. Really have to see the IL for that last bit though, could be an FSharpRef or some other hiding in there.

I think the difference is just that the state machines are bigger, and when their state gets boxed this results in more allocated heap size.  We would need to check actual number of allocations.

For example the F# state machine for `AsyncBinds` has this:
```
    .field public class [FSharp.Core]Microsoft.FSharp.Core.Unit Result
    .field public int32 ResumptionPoint
    .field public valuetype [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<class [FSharp.Core]Microsoft.FSharp.Core.Unit> MethodBuilder
    .field public valuetype [System.Runtime]System.Runtime.CompilerServices.YieldAwaitable/YieldAwaiter awaiter
    .field public valuetype [System.Runtime]System.Runtime.CompilerServices.YieldAwaitable/YieldAwaiter awaiter0
    .field public valuetype [System.Runtime]System.Runtime.CompilerServices.YieldAwaitable/YieldAwaiter awaiter1
    .field public valuetype [System.Runtime]System.Runtime.CompilerServices.YieldAwaitable/YieldAwaiter awaiter2
    .field public valuetype [System.Runtime]System.Runtime.CompilerServices.YieldAwaitable/YieldAwaiter awaiter3
    .field public valuetype [System.Runtime]System.Runtime.CompilerServices.YieldAwaitable/YieldAwaiter awaiter4
    .field public valuetype [System.Runtime]System.Runtime.CompilerServices.YieldAwaitable/YieldAwaiter awaiter5
    .field public valuetype [System.Runtime]System.Runtime.CompilerServices.YieldAwaitable/YieldAwaiter awaiter6
    .field public valuetype [System.Runtime]System.Runtime.CompilerServices.YieldAwaitable/YieldAwaiter awaiter7
    .field public valuetype [System.Runtime]System.Runtime.CompilerServices.YieldAwaitable/YieldAwaiter awaiter8
```
where the C# one has this:
```
    .field public int32 '<>1__state'
    .field public valuetype [System.Threading.Tasks]System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1<int32> '<>t__builder'
    .field private valuetype [System.Runtime]System.Runtime.CompilerServices.YieldAwaitable/YieldAwaiter '<>u__1'
```
Looking at that I'm quite surprised the difference isn't actually great, but `sizeof<MethodBuilder>` is 12 and `sizeof<System.Runtime.CompilerServices.YieldAwaitable.YieldAwaiter>` is actually just one byte (I suppose a boolean), so I think that adds up.

dsyme commented:
(I'm not immediately inclined to implement the state machine field sharing that C# has.  We don't do that optimization for sequence expressions)

dsyme commented:
The build is failing on Linux and Mac because the bootstrap compiler is not being used, see #6380

NinoFloris commented:
>I'll do a few more runs tomorrow, the results are a bit variable. I suspect the C# one just came in slow on that particular test run

Good to hear that, local machine runs are hard to get stable, thermal throttling and overactive background services cause night and day differences.

>I'm not immediately inclined to implement the state machine field sharing that C# has, partly because it can be a nightmare for debugging and hell if it goes wrong. Also we don't do that optimization for sequence expressions.

Agreed, it's a bit more work during GC to track the extra references inside most of the awaiters (`TaskAwaiter` carries the ref of its `Task`) but it's much more straight forward like this. Always have the option to re-evaluate if somebody wants to build the next Kestrel in F# ;)

That does make me remember something else C# does, which **is** valuable to add. 
Take `TaskAwaiter<'T>`, as it stores a reference to `Task<'T>` which can carry a user supplied 'state' object, continuations, and 'T which could be a huge object graph. The C# implementation sets the default value on the awaiter field once its done with it. Unrooting the Task reference and removing the possibility of pushing earlier allocations out to Gen 0/1 in longer running methods. 

Also, and this applies more to C# — we don't have `let! mutable` — if you would receive a new unshared instance of 'T from your callee you'd expect nulling out its only visible binding allows the GC to clean it up. Keeping the Task alive in the state machine  would prevent that.

forki commented:
@dsyme I often have to use this:

     task { return myID }

would love to see some nice syntactic sugar and/or optimization that doesn't even create a task

ashtonkj commented:
> @dsyme I often have to use this:
> 
> ```
>  task { return myID }
> ```
> 
> would love to see some nice syntactic sugar and/or optimization that doesn't even create a task

Couldn't you just use Task.FromResult?

forki commented:
right! Does that create an actual task as overhead?

ashtonkj commented:
> right! Does that create an actual task as overhead?

As far as I can see here (https://referencesource.microsoft.com/mscorlib/R/11a386e7d7cae64a.html) and here (https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.fromresult?view=netframework-4.8#remarks) it creates a Task with the result immediately set and the status of RanToCompletion

forki commented:
ok. question is: can we do similar in `task { }` for stuff that's basically already done?

ashtonkj commented:
> ok. question is: can we do similar in `task { }` for stuff that's basically already done?

I'm not sure I understand the question. Wouldn't `return` already do that?

forki commented:
yeah maybe I'm stupid here. need to think about my question ...

benaadams commented:
If it could be either; but usually is complete; then you can use ValueTask instead of Task.

dsyme commented:
Closing in favour of https://github.com/dotnet/fsharp/pull/6811 from a feature branch
